fix(DatePicker): prevent invalid/warn states when disabled/readonly

Tiss29 opened the pull request:
Fixed DatePicker components in both React and Web Components to not show invalid/warn states when in disabled or readonly state, as users cannot interact with these components in those states.

Closes [#20722](https://github.com/carbon-design-system/carbon/issues/20722)


Fix DatePicker components to not show invalid/warn states when disabled or readonly

### Changelog

**New**

- Added tests in DatePicker-test.js and date-picker_spec.ts to verify invalid/warn states are not shown when disabled or readonly.

**Changed**

- Added disabled prop to DatePicker.tsx and modified child component props to include conditions for invalid/warn states.
- Updated DatePickerInput.tsx to not apply invalid/warn CSS classes when disabled/readonly.
- Modified useNormalizedInputProps.tsx to check for disabled state when setting invalid/warn properties.
- Updated date-picker-input.ts to check for disabled state in normalizedProps.
- Added CSS rules in _date-picker.scss to remove invalid/warn styling for disabled and readonly inputs.
- Added CSS rules in date-picker.scss for web components implementation.



#### Testing / Reviewing

- Access DatePicker on storybook and change the state to "Invalid = true" and "disabled = true". The invalid message should not be shown.
- Repeat above test with "readOnly = true" instead of disabled. The invalid message should not be shown.
- Repeat the two tests above using "warn = true" instead of "invalid = true". The warning message on both scenarios should not be shown.

## PR Checklist


As the author of this PR, before marking ready for review, confirm you:

- [x] Reviewed every line of the diff
- [ ] Updated documentation and storybook examples
- [x] Wrote passing tests that cover this change
- [x] Addressed any impact on accessibility (a11y)
- [x] Tested for cross-browser consistency
- [x] Validated that this code is ready for review and status checks should pass

More details can be found in the [pull request guide](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/reviewing-pull-requests.md)

Tiss-Joseph committed:
fix(DatePicker): prevent invalid/warn states when disabled/readonly

Fixed DatePicker components in both React and Web Components to not show
invalid/warn states when in disabled or readonly state, as users cannot
interact with these components in those states.

Tiss29 committed:
Merge branch 'main' into DatePicker

maradwan26 commented on packages/web-components/src/components/date-picker/date-picker.scss:
```diff
@@ -77,12 +77,19 @@ $css--plex: true !default;
+    display: none;
+  }
+}
+
```
Preventing invalid/warn states to be applied with disabled/readonly should just be in the component file, the styles shouldn't need to be updated

maradwan26 commented on packages/react/src/components/DatePickerInput/DatePickerInput.tsx:
```diff
@@ -173,11 +174,22 @@ const DatePickerInput = React.forwardRef(function DatePickerInput(
+    disabled,
+    invalid: invalid === true,
+    invalidText,
+    warn: warn === true,
```
```suggestion
    invalid,
    invalidText,
    warn,
```
This should be able to work without the conditions

Tiss29 replied:
yes  I have made the changes can you check

maradwan26 reviewed (changes requested):
Thanks for contributing! Just some comments on coding conventions and best practices

Tiss29 commented on packages/react/src/components/DatePicker/DatePicker.tsx:
```diff
@@ -546,24 +552,24 @@ const DatePicker = React.forwardRef(function DatePicker(
-          warn: effectiveWarn,
+          invalid: invalid && !disabled,
+          warn: effectiveWarn && !disabled,
         });
```
yes  I have made the changes can you check

Tiss29 commented on packages/styles/scss/components/date-picker/_date-picker.scss:
```diff
@@ -91,6 +91,12 @@
+  .#{$prefix}--date-picker__input[disabled] ~ .#{$prefix}--form-requirement,
+  .#{$prefix}--date-picker__input[readonly] ~ .#{$prefix}--form-requirement {
+    display: none;
+  }
```
yes  I have made the changes can you check

Tiss-Joseph committed:
fix(DatePicker): prevent invalid/warn states when disabled/readonly

Fixed DatePicker components in both React and Web Components to not show
invalid/warn states when in disabled or readonly state, as users cannot
interact with these components in those states.

Tiss29 commented:
@maradwan26  I have a doubt regarding the requirement to “sign the DOC with both accounts.”
Could you please clarify which accounts specifically need to sign it?

Tiss29 committed:
Merge branch 'main' into DatePicker

Tiss29 committed:
Merge branch 'main' into DatePicker

Tiss-Joseph committed:
fix(date-picker): address code review comments

- Fix invalid outline showing with disabled/readonly states in Web Components
- Remove duplicate styles from shared _date-picker.scss
- Fix failing React tests by moving props to DatePickerInput
- Remove unused disabled prop from DatePicker.tsx
- Move Web Components tests to correct __tests__ directory
- Remove accidental comment from test file

Tiss29 commented on packages/react/src/components/DatePicker/DatePicker.tsx:
```diff
@@ -244,6 +244,11 @@ export interface DatePickerProps {
+   * Specify whether or not the input should be disabled
+   */
+  disabled?: boolean;
+
```
sure will remove all the disabled prop

Tiss29 committed:
Merge branch 'main' into DatePicker

Tiss29 committed:
Merge branch 'main' into DatePicker

Tiss29 commented:
@maradwan26 Could please check

maradwan26 commented on packages/web-components/src/components/date-picker/__tests__/date-picker-test.js:
```diff
@@ -0,0 +1,132 @@
+  afterEach(async () => {
+    await render(undefined, document.body);
+  });
+});
```
This might not be needed after changes are made

Tiss29 commented:
@maradwan26 in my web component action I couldn't see invalid 
<img width="1447" height="1002" alt="Screenshot 2025-11-06 at 11 36 26 PM" src="https://github.com/user-attachments/assets/9c9a746a-6929-4a84-95e4-e036d0feabc6" />
do I need to enable something while running storybook

Tiss29 committed:
Merge branch 'main' into DatePicker

Tiss-Joseph committed:
fix(date-picker): prevent invalid state when disabled/readonly

- Fix test file to use @open-wc/testing fixture pattern
- Fix invalid outline showing when disabled or readonly is true
- Add invalid/invalidText controls to Storybook stories

Tiss29 commented:
@maradwan26 Please recheck

maradwan26 commented:
@Tiss29 

Tests are failing for web components.

You can run the tests locally by running `yarn web-test-runner "src/components/date-picker/__tests__/date-picker-test.js"` in `packages/web-components`

Tiss29 committed:
Merge branch 'main' into DatePicker

Tiss-Joseph committed:
fix(date-picker): prevent invalid state when disabled/readonly

maradwan26 committed:
Merge branch 'main' into DatePicker

Tiss29 committed:
Update packages/web-components/src/components/date-picker/date-picker.scss

Co-authored-by: Mahmoud <132728978+maradwan26@users.noreply.github.com>

Tiss29 committed:
Update packages/web-components/src/components/date-picker/__tests__/date-picker-test.js

Co-authored-by: Mahmoud <132728978+maradwan26@users.noreply.github.com>

Tiss-Joseph committed:
fix(date-picker): prevent invalid state when disabled/readonly

Tiss29 commented on packages/web-components/tests/spec/date-picker_spec.ts:
```diff
@@ -76,7 +76,7 @@ describe('cds-date-picker', () => {
       datePicker = document.body.querySelector('cds-date-picker');
```
okay will revert

Tiss-Joseph committed:
fix(date-picker): prevent invalid state when disabled/readonly

Tiss29 commented:
@maradwan26 @alisonjoseph Please check

maradwan26 commented on packages/web-components/src/components/date-picker/date-picker.scss:
```diff
@@ -77,7 +77,7 @@ $css--plex: true !default;
   }
 }
-:host(#{$prefix}-date-picker-input[warn]) {
+:host(#{$prefix}-date-picker-input[warn]:not([disabled]):not([readonly])) {
```
```suggestion
:host(#{$prefix}-date-picker-input[warn]:not([invalid])) {
```

Seems the change was reverted

Tiss29 committed:
Update packages/web-components/src/components/date-picker/date-picker.scss

Co-authored-by: Mahmoud <132728978+maradwan26@users.noreply.github.com>

Tiss29 committed:
Merge branch 'main' into DatePicker

alisonjoseph committed:
chore: yarn format
